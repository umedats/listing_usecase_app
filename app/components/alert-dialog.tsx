import Button from "@mui/material/Button";
import Dialog from "@mui/material/Dialog";
import DialogActions from "@mui/material/DialogActions";
import DialogContent from "@mui/material/DialogContent";
import DialogContentText from "@mui/material/DialogContentText";
import DialogTitle from "@mui/material/DialogTitle";
import { useTranslations } from "use-intl";
import { useState } from "react";

/**
 * Props for the confirmation behavior of the alert dialog.
 */
interface ConfirmationOptions {
  /**
   * true if the alert dialog has to show confirmation behavior.
   */
  showConfirmation: boolean;
  /**
   * The content of the confirmation button.
   */
  confirmationButtonContent: string;
  /**
   * The content of the confirmation cancel button.
   */
  confirmationCancelButtonContent: string;
  /**
   * A function which is executed when a user confirms the dialog.
   */
  onConfirm: () => void;
}

/**
 * A component which alerts the user using a backdrop based Dialog.
 * @param title title of the dialog.
 * @param content content to be shown inside the dialog.
 * @param onClose a function which is executed when a user closes the dialog
 * @param confirmationOptions describe whether & how to provide a confirmation option
 * @constructor
 */
export default function AlertDialog({
  title,
  content,
  onClose,
  confirmationOptions,
}: {
  title: string;
  content: string;
  onClose?: () => void;
  confirmationOptions?: ConfirmationOptions;
}) {
  const [open, setOpen] = useState(true);
  const translations = useTranslations("AlertDialog");
  const closeButtonTranslation = confirmationOptions?.showConfirmation
    ? confirmationOptions.confirmationCancelButtonContent
    : translations("close");

  const handleClose = () => {
    setOpen(false);
    if (onClose) {
      onClose();
    }
  };

  const handleConfirm = () => {
    setOpen(false);
    confirmationOptions!.onConfirm();
  };

  return (
    <div>
      <Dialog data-testid="alertDialog" open={open} onClose={handleClose}>
        <DialogTitle>{title}</DialogTitle>
        <DialogContent>
          <DialogContentText>{content}</DialogContentText>
        </DialogContent>
        <DialogActions>
          {confirmationOptions?.showConfirmation && (
            <Button data-testid="confirmAlertDialog" onClick={handleConfirm}>
              {confirmationOptions.confirmationButtonContent}
            </Button>
          )}
          <Button
            data-testid="cancelAlertDialog"
            onClick={handleClose}
            autoFocus
          >
            {closeButtonTranslation}
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
}
