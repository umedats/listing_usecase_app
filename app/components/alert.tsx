import { Collapse, IconButton, Alert } from "@mui/material";
import CloseIcon from "@mui/icons-material/Close";
import React from "react";

/**
 * id: HTML id for the component
 * state: visibility state for the alert
 * stateHandler: function which controls the visibility of the component
 * severity: Display severity for the alert. Only supports 4 severities of Material UI Alert
 * message: Message to be displayed in the alert box
 * staticAlert: Hides the close button if set to true
 */
interface InputProps {
  id: string;
  state: boolean;
  stateHandler: any;
  severity: "error" | "warning" | "info" | "success";
  message: string;
  staticAlert?: boolean;
}

/**
 * Component to display temporary messages to user.
 * @param props
 */
export default function AlertComponent(props: InputProps) {
  return (
    <Collapse in={props.state}>
      <Alert
        data-testid={props.id}
        severity={props.severity}
        action={
          !props.staticAlert && (
            <IconButton
              aria-label="close"
              color="inherit"
              size="small"
              onClick={() => {
                props.stateHandler(false);
              }}
            >
              <CloseIcon />
            </IconButton>
          )
        }
        sx={{ marginBottom: 2 }}
      >
        {props.message}
      </Alert>
    </Collapse>
  );
}
