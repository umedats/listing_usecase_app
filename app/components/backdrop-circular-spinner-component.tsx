import { Backdrop, CircularProgress } from "@mui/material";

/**
 * Component which uses the basic Backdrop with a Circular Progress component in the foreground
 * to indicate a loading state.
 * @param showSpinner true to show the spinner.
 * @constructor
 */
export default function BackdropCircularSpinnerComponent({
  showSpinner,
}: {
  showSpinner: boolean;
}) {
  if (!showSpinner) {
    return <></>;
  }

  return (
    <Backdrop
      data-testid="backDropCircularSpinner"
      open={showSpinner}
      transitionDuration={{ appear: 0, enter: 0 }}
      sx={{ color: "#fff", zIndex: (theme) => theme.zIndex.drawer + 1 }}
    >
      <CircularProgress color="inherit" />
    </Backdrop>
  );
}
