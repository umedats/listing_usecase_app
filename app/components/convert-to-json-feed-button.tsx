import { Button } from "@mui/material";
import { useTranslations } from "use-intl";
import { cloneAndCleanupListing } from "@/app/utils/schema";
import { useContext, useState } from "react";
import { convertListingToJsonFeed } from "@/app/utils/json-feed";
import { useLocale } from "next-intl";
import { convertLocaleToSPAPIFormat } from "@/app/utils/i18n";
import { saveAs } from "file-saver";
import { serializeToJsonString } from "@/app/utils/serialization";
import AlertDialog from "@/app/components/alert-dialog";
import { SettingsContext } from "@/app/context/settings-context-provider";
import { v4 as uuid } from "uuid";
import { AlertDialogState } from "@/app/model/types";
import TooltipWrapper from "./tooltip-wrapper";

const SAVE_FILE_NAME = "JsonFeed.json";

/**
 * A button which converts the listing data entered by the user into
 * JSON_LISTINGS_FEED and saves the feed content into a file.
 * @param sku the sku of the listing.
 * @param productType the product type of the listing.
 * @param listing the listing data entered by the user.
 * @param useCase the type of the update for the listing.
 * @param initialListing the listing data before user edits.
 * @param submissionMode the submission mode to be used when the use case is update listing.
 * @constructor
 */
export default function ConvertToJsonFeed({
  sku,
  productType,
  listing,
  useCase,
  initialListing,
  submissionMode,
}: {
  sku: string;
  productType: string;
  listing: object;
  useCase: string;
  initialListing: object;
  submissionMode?: string;
}) {
  const [alertDialogState, setAlertDialogState] = useState<AlertDialogState>({
    showAlert: false,
    alertTitle: "",
    alertContent: "",
  });
  const translations = useTranslations("ConvertToJsonFeed");
  const settingsContext = useContext(SettingsContext);
  const locale = useLocale();

  const transformToJsonFeedAndDownload = () => {
    const feed = convertListingToJsonFeed(
      sku,
      productType,
      cloneAndCleanupListing(listing),
      useCase,
      initialListing,
      settingsContext.settingsState.settings.sellingPartnerId,
      convertLocaleToSPAPIFormat(locale),
      submissionMode,
    );
    if (!feed) {
      setAlertDialogState({
        showAlert: true,
        alertTitle: translations("invalidStatus"),
        alertContent: translations("invalidMessage"),
      });
      return;
    }
    const feedBlob = new Blob([serializeToJsonString(feed)], {
      type: "text/plain;charset=utf-8",
    });
    saveAs(feedBlob, SAVE_FILE_NAME);
    setAlertDialogState({
      showAlert: true,
      alertTitle: translations("successStatus"),
      alertContent: translations("successMessage", {
        fileName: SAVE_FILE_NAME,
      }),
    });
  };

  const hideAlert = () =>
    setAlertDialogState({ ...alertDialogState, showAlert: false });

  return (
    <>
      <TooltipWrapper
        title={translations("buttonHelpText")}
        placement={"right"}
      >
        <Button
          id="convertToJsonFeed"
          data-testid="convertToJsonFeed"
          variant="contained"
          onClick={transformToJsonFeedAndDownload}
        >
          {translations("buttonName")}
        </Button>
      </TooltipWrapper>
      {alertDialogState.showAlert && (
        <AlertDialog
          key={uuid()}
          title={alertDialogState.alertTitle}
          content={alertDialogState.alertContent}
          onClose={hideAlert}
        />
      )}
    </>
  );
}
