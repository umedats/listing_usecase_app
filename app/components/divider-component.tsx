import { Divider } from "@mui/material";

/**
 * Component which creates a Divider which can be used to separate logical sections in a page.
 * @constructor
 */
export function DividerComponent() {
  return (
    <Divider
      style={{
        margin: "1rem 0",
        borderBottomWidth: "3px",
        borderColor: "#333",
      }}
    />
  );
}
