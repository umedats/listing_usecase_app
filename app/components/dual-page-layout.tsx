import React from "react";
import { Grid } from "@mui/material";

/**
 * A component which displays contents of two pages in a single view using the
 * parallel routes.
 * @param props props for the component.
 * @constructor
 */
export default function DualPageLayout(props: {
  leftContent: React.ReactNode;
  rightContent: React.ReactNode;
}) {
  return (
    <Grid container spacing={1}>
      <Grid
        item
        xs={6}
        sx={{
          borderRight: "1px solid black",
          overflow: "auto",
          maxHeight: "80vh",
        }}
      >
        {props.leftContent}
      </Grid>
      <Grid
        item
        xs={6}
        sx={{
          borderLeft: "1px solid black",
          overflow: "auto",
          maxHeight: "80vh",
        }}
      >
        {props.rightContent}
      </Grid>
    </Grid>
  );
}
