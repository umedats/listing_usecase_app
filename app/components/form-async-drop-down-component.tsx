import TextField from "@mui/material/TextField";
import Autocomplete from "@mui/material/Autocomplete";
import CircularProgress from "@mui/material/CircularProgress";
import { Grid, InputLabel, Tooltip } from "@mui/material";
import useSWR, { useSWRConfig } from "swr";
import { useTranslations } from "use-intl";
import { Fragment, useState } from "react";

/**
 * Options which will be displayed in the dropdown.
 */
export interface Options {
  key: string;
  value: string;
}

/**
 * @param id: HTML identifier for this element
 * @param required: Indicates if this field is optional or not
 * @param disabled: Optional param which indicates if this field is disabled
 * @param label: Form label for this field
 * @param inputFieldLabel: Optional Label which will be displayed on the text field
 * @param noOptionsText: Optional text to be displayed when the fetcher returns empty array.
 * @param helpText: Help text for this field
 * @param onChange: Callback function to handle onChange events
 * @param query: Query value to be passed to the fetcher function
 * @param fetcher: Function which is called to load the data for the drop-down
 */
interface InputProps {
  id: string;
  required: boolean;
  disabled?: boolean;
  label: string;
  inputFieldLabel?: string;
  noOptionsText?: string;
  helpText: string;
  onChange: any;
  query: any;
  fetcher: (query: any) => Promise<Options[]>;
}

/**
 * A dropdown component which can be used to load its options dynamically based on provided context.
 * @param props
 * @constructor
 */
export default function AsyncFormDropDownComponent(props: InputProps) {
  const translations = useTranslations("AsyncFormDropDown");

  const [open, setOpen] = useState(false);

  const { data, error, isLoading } = useSWR(
    open ? props.query : null,
    props.fetcher,
  );
  const { mutate } = useSWRConfig();

  const onPopupOpen = () => {
    setOpen(true);
  };

  const onPopupClose = () => {
    setOpen(false);
    mutate(props.query, undefined, { revalidate: false });
  };

  return (
    <Grid container spacing={3} alignItems="center">
      <Grid item xs={4}>
        <Tooltip title={props.helpText} placement={"right"}>
          <InputLabel htmlFor={props.id}>{props.label}</InputLabel>
        </Tooltip>
      </Grid>
      <Grid item xs={4}>
        <Tooltip
          title={error ? translations("loadingErrorMessage") : props.helpText}
          placement={"right"}
        >
          <Autocomplete
            id={props.id}
            size="small"
            open={open}
            onOpen={onPopupOpen}
            onClose={onPopupClose}
            onChange={props.onChange}
            isOptionEqualToValue={(option, value) => option.key === value.key}
            getOptionLabel={(option) => option.key}
            noOptionsText={props.noOptionsText}
            options={data ? data : []}
            loading={isLoading}
            renderInput={(params) => (
              <TextField
                {...params}
                label={error ? translations("error") : props.inputFieldLabel}
                margin="normal"
                size="small"
                error={!!error}
                helperText={error ? translations("loadingErrorMessage") : ""}
                fullWidth
                InputProps={{
                  ...params.InputProps,
                  endAdornment: (
                    <Fragment>
                      {isLoading ? (
                        <CircularProgress color="inherit" size={20} />
                      ) : null}
                      {params.InputProps.endAdornment}
                    </Fragment>
                  ),
                }}
              />
            )}
          />
        </Tooltip>
      </Grid>
    </Grid>
  );
}
