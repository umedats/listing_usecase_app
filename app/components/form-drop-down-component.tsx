import { Grid, InputLabel, MenuItem, Select, Tooltip } from "@mui/material";

/**
 * id: HTML id for this component
 * label: Label text for the dropdown
 * helpText: Helper text which will be displayed on hover
 * selectedKey: Default value text which will be selected on render. Should be one of the elements in the options list.
 * options: List of elements displayed in dropdown. Key: unique identifier for the element. Value: Display label for the element.
 * onChange: Function to handle onchange events
 * disabled: flag to mark the drop-down as disabled.
 * xsDropDownWidth: the maximum width of the drop-down.
 */
interface InputProps {
  id: string;
  label: string;
  helpText: string;
  selectedKey: string;
  options: {
    key: string;
    label: string;
  }[];
  onChange?: any;
  disabled?: boolean;
  xsDropDownWidth?: number;
}

/**
 * Form dropdown input component with static options to be used in the forms.
 * If you need the options to be loaded dynamically based on additional context see @code AsyncFormDropDownComponent
 * @constructor
 */
export default function FormDropDownComponent(props: InputProps) {
  const dropDownWidth = props.xsDropDownWidth ?? 8;
  return (
    <Grid container spacing={3} alignItems="center">
      <Grid item xs={12 - dropDownWidth}>
        <Tooltip title={props.helpText} placement={"right"}>
          <InputLabel htmlFor={props.id}>{props.label}</InputLabel>
        </Tooltip>
      </Grid>
      <Grid item xs={dropDownWidth}>
        <Tooltip title={props.helpText} placement={"right"}>
          <Select
            id={props.id}
            data-testid={props.id}
            key={props.selectedKey}
            value={props.selectedKey}
            variant="outlined"
            size="small"
            onChange={props.onChange}
            disabled={props.disabled}
          >
            {props.options.map((option) => (
              <MenuItem id={props.id} key={option.key} value={option.key}>
                {option.label}
              </MenuItem>
            ))}
          </Select>
        </Tooltip>
      </Grid>
    </Grid>
  );
}
