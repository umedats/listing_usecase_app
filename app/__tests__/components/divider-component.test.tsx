import { render } from "@testing-library/react";
import React from "react";
import { DividerComponent } from "@/app/components/divider-component";

test("Component renders", () => {
  const { asFragment } = render(<DividerComponent />);
  expect(asFragment()).toMatchSnapshot();
});
