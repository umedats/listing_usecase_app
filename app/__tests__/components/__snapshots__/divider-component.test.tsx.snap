// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders 1`] = `
<DocumentFragment>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    style="margin: 1rem 0px; border-bottom-width: 3px; border-color: #333;"
  />
</DocumentFragment>
`;
